viomem/sys/Device.c

```diff
@@ -0,0 +1,530 @@
+
+    status = ViomemInit(Device);
+    if(!NT_SUCCESS(status))
+    {
+        TraceEvents(TRACE_LEVEL_ERROR, DBG_PNP,
+           "ViomemInit failed with status 0x%08x\n", status);
+        ViomemTerminate(Device);
+        return status;
+    }
+
+    status = ViomemCreateWorkerThread(Device);
+    if(!NT_SUCCESS(status))
+    {
+        TraceEvents(TRACE_LEVEL_ERROR, DBG_PNP,
+           "ViomemCreateWorkerThread failed with status 0x%08x\n", status);
+    } 
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Return \n", __FUNCTION__);
+
+    return status;
+}
+
+NTSTATUS 
+ViomemEvtDeviceD0Exit(
+    IN  WDFDEVICE Device,
+    IN  WDF_POWER_DEVICE_STATE TargetState
+    )
+{
+    PDEVICE_CONTEXT devCtx = GetDeviceContext(Device);
+
+    UNREFERENCED_PARAMETER(TargetState);
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Entry\n", __FUNCTION__);
+
+    PAGED_CODE();
+
+    ViomemTerminate(Device);
+	
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Return\n", __FUNCTION__);
+
+    return STATUS_SUCCESS;
+}
+
+NTSTATUS 
+ViomemEvtDeviceD0ExitPreInterruptsDisabled(
+	IN  WDFDEVICE Device,
+    IN  WDF_POWER_DEVICE_STATE TargetState
+    )
+{
+    PDEVICE_CONTEXT       devCtx = GetDeviceContext(Device);
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Entry\n", __FUNCTION__);
+
+    PAGED_CODE();
+
+    ViomemCloseWorkerThread(Device);
+    if (TargetState == WdfPowerDeviceD3Final)
+    {
+		// Nothing here ?
+	}
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Return\n", __FUNCTION__);
+
+    return STATUS_SUCCESS;
+}
+
+BOOLEAN 
+ViomemISR(
+    IN WDFINTERRUPT WdfInterrupt,
+    IN ULONG        MessageID
+    )
+{
+    PDEVICE_CONTEXT     devCtx = NULL;
+    WDFDEVICE           Device;
+    UNREFERENCED_PARAMETER( MessageID );
+
+    Device = WdfInterruptGetDevice(WdfInterrupt);
+    devCtx = GetDeviceContext(Device);
+
+    if (VirtIOWdfGetISRStatus(&devCtx->VDevice) > 0)
+    {
+        TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INTERRUPT, "%s queueing DPC\n", __FUNCTION__);
+        WdfInterruptQueueDpcForIsr( WdfInterrupt );
+        return TRUE;
+    }
+    else
+    {
+        TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INTERRUPT, "%s No ISR indicated\n", __FUNCTION__);
+    }
+    return FALSE;
+}
+
+VOID 
+ViomemDPC(
+    IN WDFINTERRUPT WdfInterrupt,
+    IN WDFOBJECT    WdfDevice
+    )
+{
+    unsigned int          len;
+    PDEVICE_CONTEXT       devCtx = GetDeviceContext(WdfDevice);
+    PVOID                 buffer;
+
+    BOOLEAN               bHostAck = FALSE;
+    UNREFERENCED_PARAMETER( WdfInterrupt );
+
+	TraceEvents(TRACE_LEVEL_INFORMATION, DBG_INIT, "%s Entry\n", __FUNCTION__);
+
+    WdfSpinLockAcquire(devCtx->infVirtQueueLock);
+    
+	if (virtqueue_has_buf(devCtx->infVirtQueue))
+	{
+        bHostAck = TRUE;
+    }
+    WdfSpinLockRelease(devCtx->infVirtQueueLock);
+
+    if(bHostAck)
+    {
+        KeSetEvent (&devCtx->hostAcknowledge, EVENT_INCREMENT, FALSE);

```

Comment:
Probably should be also called during device removal as well. 
Also for WakeUpThread event.

Might be a good idea to review event handling in a meeting.